### ISSUE TITLE
feat: Environment, as a more maintainable alternative to a bunch of envvars

### ISSUE THREAD

**Issue by mumoshu** [roboll/helmfile#253]
Extracted from https://github.com/roboll/helmfile/issues/59#issuecomment-415690069

Let's add `environments` to helmfile.yaml:

```yaml
environments:
  default:
    values:
    - ./default.0.yaml
    - ./default.1.yaml.gotmpl
  # note: values from the default env. aren't inherited to production. Use yaml anchors for inheritance.
  production:
    values:
    - ./production.yaml

releases:
- name: myapp
   chart: mychart
   values:
  - values.yaml.gotmpl
```

In the above example, `default.0.yaml` and `default.1.yaml.gotmpl` is merged into a golang map, that can be referenced by `{{ .Environment.Values.foo.bar }}` from within the values.yaml template(#216) `values.yaml.gotmpl`.

`helmfile sync` against a missing environment should [SUCCEED with a big warning message](https://github.com/roboll/helmfile/issues/59#issuecomment-416500471).

**Comment by sstarcher** [comment]
@mumoshu if by default we are going to succeed with a missing environment could we have a flag that forces it to fail if the environment is missing.

**Comment by mumoshu** [comment]
@sstarcher Thanks for the comment!

I was reconsidering to make the environment required, whenever it isn't `default`. That is `helmfile sync` doesn't fail or wan, while `helmfile --environment prod sync` fails when helmfile.yaml misses the prod env, that should indicate the user misspelled the environment name either in the flag or the config file.

What would be your use-case though?

**Comment by sstarcher** [comment]
@mumoshu after thinking through it I have already changed my mind.

My specific use-case I would have some helmfiles that contain a single specific environment and if that environment only had production contents I would certainly not want it to fail if someone used a different environment.

I guess overall I want the environment they specified to exist in at least one of the helmfiles.  If that makes sense.

**Comment by mumoshu** [comment]
@sstarcher Thanks for the response!

So, you're saying that `helmfile --environment production -f helmfile.d/ sync` should fail only when there's no helmfile.yaml that contains the `production` environment, correct?

The behavior seems similar to that of `helmfile --selector foo=bar sync`, which makes it sounds good to me.

**Comment by mumoshu** [comment]
The change discussed above has been implemented in #294
